### PR TITLE
[DRAFT] Add instance-uuid->job-uuid cache and update kubernetes/controller tests

### PR DESCRIPTION
**Pull request by calebhar12**
## Changes proposed in this PR

- Add a cache mapping `instance-uuid` to `job-uuid`
- Update `assoc-uuids` function to get `job-uuid`s from cache when needed and append them to passport event maps
- Add `with-redefs` logic to spoof cache hits in `cook.test.kubernetes.controller` tests

## Why are we making these changes?

- Adding this cache gives us the ability to log `job-uuid` for all Passport events in the state machine where `job-uuid` is not immediately visible
- These changes will give us a way to associate `instance-uuid` with `job-uuid` without querying datomic each time we want to access it

### CODE REVIEW

File: scheduler/src/cook/caches.clj [hunk 1]
```diff
@@ -40,3 +40,8 @@
 (mount/defstate ^Cache pool-name->accepts-submissions?-cache :start (new-cache config/config))
 (mount/defstate ^Cache pool-name->db-id-cache :start (new-cache config/config))
 (mount/defstate ^Cache user-and-pool-name->quota :start (new-cache config/config))
+(mount/defstate ^Cache instance-uuid->job-uuid :start
+  (-> (CacheBuilder/newBuilder)
+    (.maximumSize (get-in config/config [:settings :cache-working-set-size])) ;1million
```
Comment (dposada) [inline scheduler/src/cook/caches.clj:45]:
```suggestion
    (.maximumSize (get-in config/config [:settings :cache-working-set-size]))
```

File: scheduler/src/cook/kubernetes/api.clj [hunk 2]
```diff
@@ -82,11 +82,14 @@
...
-  "Return a map with :job-uuid or :instance-uuid based on pod-name"
+(defn assoc-uuids
+  "Returns event-map with added fields :job-uuid (based on pod-name) and :instance-uuid (if pod is not synthetic)"
   [event-map pod-name]
   (let [instance-uuid (pod-name->instance-uuid pod-name)
-        job-uuid (pod-name->job-uuid pod-name)]
+        synthetic-pod-job-uuid (synthetic-pod-name->job-uuid pod-name)
+        job-uuid (if synthetic-pod-job-uuid
+                   synthetic-pod-job-uuid
+                   (tools/instance-uuid->job-uuid-cache-lookup instance-uuid))]
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/api.clj:92]: Can be simplified to:

```suggestion
        job-uuid (or synthetic-pod-job-uuid
                   (tools/instance-uuid->job-uuid-cache-lookup instance-uuid))]
```

File: scheduler/src/cook/kubernetes/compute_cluster.clj [hunk 3]
```diff
@@ -290,6 +293,8 @@
         ^V1Pod pod (api/task-metadata->pod pod-namespace compute-cluster task-metadata)
         new-cook-expected-state-dict {:cook-expected-state :cook-expected-state/starting
                                       :launch-pod {:pod pod}}]
+    (when (api/synthetic-pod? pod-name)
+      (cache/put-cache! caches/instance-uuid->job-uuid identity pod-name (get-in task-metadata [:task-request :job :job/uuid])))
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/compute_cluster.clj:297]: I think it'd be nice to include a comment explaining why this is needed / giving some context for it.

File: scheduler/src/cook/rest/api.clj [hunk 4]
```diff
@@ -1438,7 +1429,7 @@
   (let [jobs (wrap-seq (get-in ctx [:request :query-params :job]))
         instances (wrap-seq (get-in ctx [:request :query-params :instance]))
         allow-partial-results (get-in ctx [:request :query-params :partial])
-        instance-uuid->job-uuid #(instance-uuid->job-uuid (d/db conn) %)
+        instance-uuid->job-uuid #(util/instance-uuid->job-uuid (d/db conn) %)
```
Comment (dposada) [inline scheduler/src/cook/rest/api.clj:1432]: Would it make sense to use the cached call here?

File: scheduler/src/cook/rest/api.clj [hunk 5]
```diff
@@ -1490,7 +1481,7 @@
   cluster, and get back the data for those that do match."
   (let [uuids (wrap-seq (::instances ctx))
         allow-partial-results? (::allow-partial-results? ctx)
-        instance-uuid->job-uuid #(instance-uuid->job-uuid (d/db conn) %)
+        instance-uuid->job-uuid #(util/instance-uuid->job-uuid (d/db conn) %)
```
Comment (dposada) [inline scheduler/src/cook/rest/api.clj:1484]: Would it make sense to use the cached call here?

File: scheduler/src/cook/caches.clj (hunk 1 above)
Comment (calebhar12) [reply scheduler/src/cook/caches.clj:45]: Good catch, will fix

File: scheduler/src/cook/config.clj [hunk 6]
```diff
@@ -512,7 +512,8 @@
                                   :default-pool "no-pool"}
                                  pool-selection)})))
      :passport (fnk [[:config {passport {}}]]
-                 passport)
+                 (merge passport
+                        {:instance-uuid->job-uuid-cache-expiry-time-hours 24}))
```
[the remark this answers is not in the document]
Comment (calebhar12) [reply scheduler/src/cook/config.clj:516]: Will do.

File: scheduler/src/cook/kubernetes/api.clj [hunk 7]
```diff
@@ -70,7 +70,7 @@
   [pod-name]
   (str/starts-with? pod-name cook-synthetic-pod-name-prefix))
 
-(defn pod-name->job-uuid
+(defn synthetic-pod-name->job-uuid
```
[the remark this answers is not in the document]
Comment (calebhar12) [reply scheduler/src/cook/kubernetes/api.clj:73]: Yeah I see the implication, I'll change it back.

File: scheduler/src/cook/kubernetes/api.clj (hunk 2 above)
Comment (calebhar12) [reply scheduler/src/cook/kubernetes/api.clj:92]: Yeah I'll definitely use this. I thought Clojure would have a sleeker way to do it than my solution but I didn't know `or` worked like this, very interesting.

File: scheduler/src/cook/kubernetes/compute_cluster.clj (hunk 3 above)
Comment (calebhar12) [reply scheduler/src/cook/kubernetes/compute_cluster.clj:297]: Do you mean specifically the `(when (api/synthetic-pod? pod-name)` part? Or do you just mean to provide context on why the cache exists.

File: scheduler/src/cook/rest/api.clj (hunk 4 above)
Comment (calebhar12) [reply scheduler/src/cook/rest/api.clj:1432]: In short, yes it can be.

My only concern is that currently we cache the `instance-uuid-->job-uuid` information right before the task is pumped into the k8s state machine, which I think is after that information is added to datomic (I'm not 100% sure). So if this `rest.api` function is called in that interim period before adding `instance-uuid` to the cache but after adding it to the DB, our cache would miss. I'm not super familiar with how often this `retrieve-jobs` api call is generally used, but I assume that it isn't used frequently enough in that interim period to give our cache a significantly increased miss rate.

Comment (calebhar12) [reply scheduler/src/cook/rest/api.clj:1432]: On the subject of this time difference-- Could you point me to where the `instance-uuid-->job-uuid` information is added to datomic? I thought it might be in https://github.com/twosigma/Cook/blob/master/scheduler/src/cook/scheduler/scheduler.clj#L773 where `:job/_instance` is in the map but I wasn't 100% sure if that's what's happening in that function (not too familiar with how datomic+clojure works). I'm interested  to see how early it occurs vs. the time when we cache it.

File: scheduler/src/cook/rest/api.clj (hunk 5 above)
Comment (calebhar12) [reply scheduler/src/cook/rest/api.clj:1484]: (see above comment, same concept)

File: scheduler/src/cook/kubernetes/compute_cluster.clj (hunk 3 above)
Comment (dposada) [reply scheduler/src/cook/kubernetes/compute_cluster.clj:297]: I meant context on why the cache exists (but maybe the synthetic pod check warrants a little explanation too)

File: scheduler/src/cook/tools.clj [hunk 8]
```diff
@@ -1037,3 +1038,22 @@
...
+
+(defn instance-uuid->job-uuid-cache-miss
+  "Wrapper function that passes in a DB to the instance-uuid->job-uuid function. Used for cache misses by the instance-uuid->job-uuid cache"
+  [instance-uuid]
+  (instance-uuid->job-uuid (d/db datomic/conn) instance-uuid))
+
+(defn instance-uuid->job-uuid-cache-lookup
+  "Gets value from cache if it is present, else search datomic for it"
+  [instance-uuid]
+  (ccache/lookup-cache! caches/instance-uuid->job-uuid identity instance-uuid->job-uuid-cache-miss instance-uuid))
```
Comment (calebhar12) [inline scheduler/src/cook/tools.clj:1059]: @nsinkov We discussed moving the bottom two functions, `instance-uuid->job-uuid-cache-lookup` and `instance-uuid->job-uuid-cache-miss` to the cook.cached-queries file and leaving `instance-uuid->job-uuid` in cook.tools. However, the body of `instance-uuid->job-uuid-cache-miss` calls `instance-uuid->job-uuid`, so making this change resulted in a circular dependency.

Let me know if I misunderstood what we discussed, or if there's another location that you want to move these functions to.

**Review by dposada — APPROVED** [review]

File: scheduler/src/cook/rest/api.clj [hunk 9]
```diff
@@ -1438,7 +1429,7 @@
   (let [jobs (wrap-seq (get-in ctx [:request :query-params :job]))
         instances (wrap-seq (get-in ctx [:request :query-params :instance]))
         allow-partial-results (get-in ctx [:request :query-params :partial])
-        instance-uuid->job-uuid #(instance-uuid->job-uuid (d/db conn) %)
+        instance-uuid->job-uuid #(cached-queries/instance-uuid->job-uuid (d/db conn) %)
```
Comment (dposada) [inline scheduler/src/cook/rest/api.clj:1432]: Doesn't this namespace kind of imply that you're using a query that has caching (while this particular function doesn't actually have any caching)?

File: scheduler/src/cook/tools.clj [hunk 10]
```diff
@@ -28,6 +28,7 @@
             [cook.cached-queries :as cached-queries]
             [cook.caches :as caches]
             [cook.config :as config]
+            [cook.datomic :as datomic]
```
Comment (dposada) [inline scheduler/src/cook/tools.clj:31]: Why was this needed?

Comment (calebhar12) [reply scheduler/src/cook/tools.clj:31]: Sorry that's my fault, I added the requirement for the cache functions but forgot to take it out after removing them. I'll remove it in the next revision.

File: scheduler/src/cook/rest/api.clj (hunk 9 above)
Comment (calebhar12) [reply scheduler/src/cook/rest/api.clj:1432]: Yeah, I see what you mean. Do you have ideas for what would be a better namespace for the `instance-uuid->job-uuid` function? Keeping it in `cook.rest.api` or `cook.tools` (which we originally wanted to do) resulted in circular dependency issues with `cached-queries`. And since we really want to keep `instance-uuid->job-uuid-cache-lookup` in `cached-queries`, we decided that `cached-queries` was the best namespace for `instance-uuid->job-uuid` as well since it resolves the dependency issues and makes clear the relationship between `instance-uuid->job-uuid-cache-lookup` and `instance-uuid->job-uuid`.

Comment (dposada) [reply scheduler/src/cook/rest/api.clj:1432]: I'm not seeing a good namespace that already exists.

Maybe this could be solved with function naming? Like maybe the cached one should be named `instance-uuid->job-uuid` and the one with no caching should be named `instance-uuid->job-uuid-no-caching` (or something similar)?

Comment (calebhar12) [reply scheduler/src/cook/rest/api.clj:1432]: That makes sense-- we could also maybe do something like `instance-uuid->job-uuid-datomic-query` for the `no-caching` call. The calls would then be:

`cached-queries/instance-uuid->job-uuid` when using the cache
`cached-queries/instance-uuid->job-uuid-no-caching` or `cached-queries/instance-uuid->job-uuid-datomic-query` when using datomic.

That makes what both calls do be pretty clear. Let me know if you have a preference, and I'll change it to that in the next revision.

Comment (dposada) [reply scheduler/src/cook/rest/api.clj:1432]: Sounds great, I like the `-datomic-query` idea

File: scheduler/src/cook/cached_queries.clj [hunk 11]
```diff
@@ -16,3 +19,20 @@
...
+   Returns nil if the instance uuid doesn't correspond
+   to a job"
+  [db instance-uuid]
+  (->> (d/entity db [:instance/task-id (str instance-uuid)])
+       :job/_instance
+       :job/uuid))
+
+(let [miss-fn
+      (fn [instance-uuid]
+        (instance-uuid->job-uuid-datomic-query (d/db datomic/conn) instance-uuid))]
```
Comment (dposada) [inline scheduler/src/cook/cached_queries.clj:34]: Nitpick (feel free to ignore):

I think you could instead use [`partial`](https://clojuredocs.org/clojure.core/partial):

```suggestion
      (partial instance-uuid->job-uuid-datomic-query (d/db datomic/conn))]
```

File: scheduler/src/cook/cached_queries.clj [hunk 12]
```diff
@@ -16,3 +19,20 @@
...
+  (defn instance-uuid->job-uuid-cache-lookup
```
Comment (dposada) [inline scheduler/src/cook/cached_queries.clj:35]: Still planning on removing the `-cache-lookup` from this one?

File: scheduler/src/cook/cached_queries.clj (hunk 11 above)
Comment (calebhar12) [reply scheduler/src/cook/cached_queries.clj:34]: I think I'll stick with defining the function as-is for now because it's a bit more readable to me. We would also only ever call the partial function with exactly 1 additional argument so I don't think it's necessary to use `partial`.

(With that said I'm still new to Clojure, so if it's a highly recommended convention to use partial just let me know.)

File: scheduler/src/cook/cached_queries.clj (hunk 12 above)
Comment (calebhar12) [reply scheduler/src/cook/cached_queries.clj:35]: I decided not to because I want it to be clear that this particular function getting `job-uuid` from `instance-uuid` is looking in a cache as opposed to looking straight t datomic. And for that, I thought it made more sense to have:

`instance-uuid->job-uuid-datomic-query` & `instance-uuid->job-uuid-cache-lookup`

vs.

`instance-uuid->job-uuid-datomic-query` & `instance-uuid->job-uuid`

Imo, the former makes it more clear what each function does. Thoughts?

File: scheduler/src/cook/cached_queries.clj (hunk 11 above)
Comment (dposada) [reply scheduler/src/cook/cached_queries.clj:34]: Sounds good, I think either way is fine

File: scheduler/src/cook/cached_queries.clj (hunk 12 above)
Comment (dposada) [reply scheduler/src/cook/cached_queries.clj:35]: I don't feel too strongly one way or the other. 

I definitely see your point, but it's not consistent with the other two existing functions in `cached-queries` (`job->pool-name` and `job-ent->user`). 

I'm fine with either.

**Review by dposada — APPROVED** [review]

**Review by nsinkov — APPROVED** [review]
